ros_ws/src/teleoperation/src/keyboard_controller.cpp

```diff
@@ -1,73 +1,126 @@
+}
 
-        switch (key)
+void KeyboardController::pollKeyboardEvents()
+{
+    SDL_Event event;
+    while (SDL_PollEvent(&event))
+    {
+        if (event.type == SDL_KEYUP || event.type == SDL_KEYDOWN)
         {
-            case Keycode::W:
-                velocity += 1;
-                break;
-            case Keycode::S:
-                velocity -= 1;
-                break;
-            case Keycode::A:
-                angle += 1;
-                break;
-            case Keycode::D:
-                angle -= 1;
-                break;
+            for (int i = 0; i < KEY_COUNT; i++)
+            {
+                if ((uint16_t)event.key.keysym.sym == (int)KEY_CODES[i])
+                {
+                    this->m_key_pressed_state[i] = event.type == SDL_KEYDOWN;
+                }
+            }
         }
+        else if (event.type == SDL_QUIT)
+        {
+            ros::shutdown();
+            exit(EXIT_SUCCESS);
+        }
+    }
+}
 
-        this->publishDriveParameters(velocity, angle);
+void KeyboardController::keyboardLoop()
+{
+    ros::Rate timer(PARAMETER_UPDATE_FREQUENCY);
+    double delta_time = timer.expectedCycleTime().toSec();
+
+    while (ros::ok())
+    {
+        this->pollKeyboardEvents();
+        this->updateDriveParameters(delta_time);
+        this->publishDriveParameters();
+
+        ros::spinOnce();
+        timer.sleep();
     }
 }
 
-int KeyboardController::getKeyboardCharacter()
+void KeyboardController::updateDriveParameters(double delta_time)
 {
-    static struct termios old_terminal, new_terminal;
-    // back up current terminal settings
-    tcgetattr(STDIN_FILENO, &old_terminal);
-    new_terminal = old_terminal;
-    // disable buffering
-    new_terminal.c_lflag &= ~ICANON;
-    // apply new settings
-    tcsetattr(STDIN_FILENO, TCSANOW, &new_terminal);
-
-    // read character (non-blocking)
-    int character = getchar();
-
-    // restore old settings
-    tcsetattr(STDIN_FILENO, TCSANOW, &old_terminal);
-
-    return character;
+    double steer = this->m_key_pressed_state[(int)KeyIndex::STEER_LEFT]
+        ? +1
+        : (this->m_key_pressed_state[(int)KeyIndex::STEER_RIGHT] ? -1 : 0);
+    double throttle = this->m_key_pressed_state[(int)KeyIndex::ACCELERATE]
+        ? +1
+        : (this->m_key_pressed_state[(int)KeyIndex::DECELERATE] ? -1 : 0);
+
+    double steer_limit = map(fabs(this->m_velocity), 0, MAX_VELOCITY, 1, FAST_STEER_LIMIT);
+    double angle_update = steer * delta_time * STEERING_SPEED;
+    this->m_angle = clamp(this->m_angle + angle_update, -MAX_STEERING * steer_limit, +MAX_STEERING * steer_limit);
+    double velocity_update = throttle * delta_time * (this->m_velocity * throttle > 0 ? ACCELERATION : BRAKING);
+    this->m_velocity = clamp(this->m_velocity + velocity_update, -MAX_VELOCITY, +MAX_VELOCITY);
+
+    if (steer == 0 && this->m_angle != 0)
+    {
+        double sign = this->m_angle > 0 ? 1 : -1;
```

Comment:
You could replace this with 
```C++
auto sign = copysign(1.0, m_angle);
```
See https://en.cppreference.com/w/cpp/numeric/math/copysign for more information